### PR TITLE
GRAM: improve parser recovering in several cases

### PR DESCRIPTION
**Pull request by vlad20012**
1. Better recover on unpaired '}' brace, e.g.
```rust
fn foo(){}
}
fn bar(){}
```
2. better struct fields recovery
```rust
struct S2 {
    f1: i32 // missing comma
    f2: i32
}

struct S3 {
    : i32 // missing identifier and comma
    f2: i32
}
```
3. better tuple struct fields recovery
```rust
struct S4(i32 i32);

struct S5(, i32);
```
4. better struct literal fields recovery
```rust
fn main() {
    let a = S { , a: 1 } // may starts with something invalid
}
```
5. better match expr arms recovery
```rust
fn main() {
    match x {
        => {} // A first arm now may starts with something invalid
        _ => {}
    }
}
```
6. better function parameters recovery
```rust
fn foo(, b: i32) {} // Function parameters now may start with something invalid
fn bar( {} // recover on missed `)`
```
7. better field, tuple & slice pattern recovery
```rust
fn main() {
    let S {,f} = S { f: 92.0 };

    let (a,, b);
    let (, c);

    let [a,, b];
    let [, c];
}
```

### CODE REVIEW

**Comment by vlad20012** [comment]
This actually breaks tests for `RsMainFunctionNotFoundInspection` hack (#5547)

**Comment by Kobzol** [comment]
> This actually breaks tests for `RsMainFunctionNotFoundInspection` hack (#5547)

Haha, that was the first thing that came to my mind when I saw this PR :D I suppose that the "hack" might still be useful, we should just find a different parser-breaking code for the test? Or is it impossible to break the parser on the top file level with this pR?

**Comment by vlad20012** [comment]
>  Or is it impossible to break the parser on the top file level with this pR?

I don't know 😂
Looks like recovering on `}` is the most powerful thing, and it's not easy now to break the entire file parsing

**Review by Undin — APPROVED** [review]
Great!
Thank you!

File: src/main/grammars/RustParser.bnf [hunk 1]
```diff
@@ -734,19 +742,30 @@ EnumVariant ::= OuterAttr* identifier VariantArgs? {
...
   pin = 1
   extends = "org.rust.lang.core.psi.ext.RsStubbedElementImpl<?>"
   stubClass = "org.rust.lang.core.stubs.RsPlaceholderStub"
   elementTypeFactory = "org.rust.lang.core.stubs.StubImplementationsKt.factory"
 }
+private TupleFieldDecl_with_recover ::= !')' TupleFieldDecl (',' | &')') {
+  pin = 1
+  recoverWhile = TupleFieldDecl_recover
+}
+private TupleFieldDecl_recover ::= !(')' | AttrsAndVis_first | TypeReference_first)
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/grammars/RustParser.bnf:782]: I added more tests

**Comment by vlad20012** [comment]
bors r=undin

**Pull request merged**